### PR TITLE
Adversarial unit test sweep: repair rotted tests, harden server against malformed payloads

### PR DESCRIPTION
**Pull request by yjsoon**
## Summary

Comprehensive adversarial unit test sweep targeting edge cases, malformed inputs, boundary values, resource exhaustion, and state corruption. Final state: **169 server unit tests + 124 teacher tests**, all green, plus live-server integration checks, typecheck, and production build verified.

### Rotted tests repaired
- `ActivityRoom.test.js` and `activityHandler.test.js` used vitest imports but the server runs `node:test` — they failed to even load, and the `test:unit` glob missed their directories so this was invisible. Converted to `node:test` and widened the glob to `src/**/*.test.js`.

### Server bugs found by the sweep (each with a regression test)
- **Crash vectors**: `ActivityRoom.evaluateAnswers` threw on malformed student payloads (`answers: null`, non-array `placements`, non-string text inputs, inherited object keys). With no global exception handling, a throwing socket handler kills the whole server. Hardened the room, normalized handler payloads, and added an `installSafeSocketEvents` wrapper so any handler throw logs + acks `INTERNAL_ERROR` instead of crashing.
- **Resource exhaustion**: `MAX_QUESTIONS_PER_ROOM`, `MAX_SUBMISSIONS_PER_ROOM`, and `MAX_SESSIONS` were defined but never enforced. Now enforced. `session:join` also stored unbounded client-supplied names/studentIds — now capped server-side.
- **State corruption**: `PollRoom.vote('constructor')` passed the `!== undefined` check via the prototype chain and corrupted vote totals to NaN. `RETRY_NOT_ALLOWED`/`SUBMISSION_FAILED` error codes were used but undefined, silently degrading to `INTERNAL_ERROR`. `Session.getActiveRooms` truncated widget IDs containing `:`.

### Teacher bugs found by the sweep
- The FillBlank/CodeFillBlank tests validated **in-file copies** of the activity-builder logic that had diverged from the shipping code (different blank-marker syntax). Extracted the real builders into `activityBuilders.ts` and pointed tests at them.
- That surfaced a real scoring bug: duplicate answers (same word in two blanks, or an answer duplicated as a distractor) pointed every blank at the first copy via `indexOf`, marking identical dragged words wrong. Targets now accept every item with matching content.
- `generateRandomFood` treated duplicate snake segments as a full grid.

### Test audit
- Strengthened tautological tests (Card asserted only static class strings; enableRenderTracking asserted only `not.toThrow()`).
- New adversarial coverage: validation boundaries, rate-limiter windows (mocked clocks), code-generator exhaustion, wordle game hook, session join hostile payloads.

## Test plan
- [x] `npm run test:unit` (server) — 169 tests pass
- [x] `npm test` (teacher, vitest) — 124 tests pass
- [x] `node test-server.js` against a live server — 11/11 pass
- [x] `tsc --noEmit` and `vite build` (teacher) clean

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_014XLfbpRxnHs5fx72nZsfwb

---
_Generated by [Claude Code](https://claude.ai/code/session_014XLfbpRxnHs5fx72nZsfwb)_

---

<a href="https://app.devin.ai/review/tinkertanker/classroom-widgets/pull/67" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

File: packages/server/package.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
     "dev": "nodemon src/server.js",
     "test": "node scripts/check-startup-warnings.js && npm run test:unit && node test-server.js",
     "test:startup": "node scripts/check-startup-warnings.js",
-    "test:unit": "node --test src/*.test.js src/routes/*.test.js src/services/*.test.js"
+    "test:unit": "node --test \"src/**/*.test.js\""
```
[the remark this answers is not in the document]
Comment (yjsoon) [reply packages/server/package.json:13]: Good catch on the version dependency — addressed in ca8ae9c by adding `"engines": { "node": ">=22" }` to the server package, so npm warns clearly on older runtimes.

One correction to the failure-mode analysis: on Node <21 the quoted glob doesn't silently match nothing — `node --test` treats it as a literal path and fails loudly with `Cannot find module` (exit 1), so tests can't silently pass. Also worth noting the repo's Dockerfiles (node:20) never run tests (build-only), so the glob only affects dev machines and future CI.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/server/src/sockets/safeSocketEvents.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * If the client supplied an acknowledgement callback, it receives a
+ * standardized INTERNAL_ERROR response instead of silence.
+ */
+function installSafeSocketEvents(socket) {
+  const rawOn = socket.on.bind(socket);
+
+  socket.on = (event, handler) => rawOn(event, (...args) => {
+    try {
+      return handler(...args);
```
[the remark this answers is not in the document]
Comment (yjsoon) [reply packages/server/src/sockets/safeSocketEvents.js]: Correct — async handlers rejected right past the try/catch. Fixed in a0cd4b3: the wrapper now attaches a rejection handler to promise-like returns, so async failures go through the same logging + INTERNAL_ERROR ack path, with regression tests for both the rejecting and resolving async cases.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/server/src/sockets/handlers/linkShareHandler.js [hunk 3]
```diff
@@ -105,6 +105,11 @@ module.exports = function linkShareHandler(io, socket, sessionManager, getCurren
 
     const submission = room.addSubmission(studentName, content, isLink);
 
+    if (!submission) {
+      socket.emit(EVENTS.LINK_SHARE.SUBMITTED, createErrorResponse('ROOM_FULL'));
```
[the remark this answers is not in the document]
Comment (yjsoon) [reply packages/server/src/sockets/handlers/linkShareHandler.js:109]: Real client bug, though it predates this branch — all 12 pre-existing error paths on this event (INVALID_SESSION, WIDGET_PAUSED, etc.) already emit the standardized `{ error: { code, message } }` shape, so ROOM_FULL is consistent with them rather than the first offender.

Fixed on the client side in 8cdb8f8: LinkShareActivity now unwraps `.message` from object-shaped errors (same dual-format handling QuestionsActivity already uses), keeping the legacy string shape working. Rebuilt student bundle included since build output is tracked.

---
_Generated by [Claude Code](https://claude.ai/code)_
